### PR TITLE
[SYCL] Pass driver options to JIT compiler and linker

### PR DESCRIPTION
**Pull request by HabKaffee**
This patch implements passing compiler and link options to sycl::compile() and sycl::link() respectively.

### CODE REVIEW

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 1]
```diff
@@ -1618,10 +1618,13 @@ ProgramManager::compile(const device_image_plain &DeviceImage,
   // TODO: Set spec constatns here.
 
   // TODO: Handle zero sized Device list.
+
+  const RTDeviceBinaryImage *ImgPtr = InputImpl->get_bin_image_ref();
```
Comment (cperkinsintel) [inline sycl/source/detail/program_manager/program_manager.cpp:1622]: maybe just a matter of style, but mightn't `const char *compileOptions = InputImpl->get_bin_image_ref()->getCompileOptions();` be slightly better than this juggling of `*ImgPtr` and `&Img` ?    The `Img` var is only used one place (though there is another one of these jugglings down in the loop in 1660

**Review by cperkinsintel — DISMISSED** [review]
LGTM

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 1 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1622]: Yes, I agree with you. I'll change it.

**Comment by HabKaffee** [comment]
I found, that this patch is affecting 1 more test in intel/llvm-test-suite. Working on it.

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
Some checks is failing because they have workarounds to "pass" until this patch will be merged. https://github.com/intel/llvm-test-suite/pull/806 exclude these workarounds and make tests work properly.

**Comment by HabKaffee** [comment]
@alexbatashev @romanovvlad ping

**Comment by HabKaffee** [comment]
@cperkinsintel ping

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 2]
```diff
@@ -1663,15 +1665,27 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
   for (const device &Dev : Devs)
     PIDevices.push_back(getSyclObjImpl(Dev)->getHandleRef());
 
+  std::vector<const char *> linkOptions;
+  for (auto &DeviceImage : DeviceImages) {
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1669]:
```suggestion
  for (const device_image_plain &DeviceImage : DeviceImages) {
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 3]
```diff
@@ -1663,15 +1665,27 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
   for (const device &Dev : Devs)
     PIDevices.push_back(getSyclObjImpl(Dev)->getHandleRef());
 
+  std::vector<const char *> linkOptions;
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1668]:
```suggestion
  std::vector<const char *> LinkOptions;
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 4]
```diff
@@ -1663,15 +1665,27 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
...
+  for (auto &DeviceImage : DeviceImages) {
+    const std::shared_ptr<device_image_impl> &InputImpl =
+        getSyclObjImpl(DeviceImage);
+    linkOptions.push_back(InputImpl->get_bin_image_ref()->getLinkOptions());
+  }
+  std::string linkOptionsStr;
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1674]:
```suggestion
  std::string LinkOptionsStr;
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 5]
```diff
@@ -1663,15 +1665,27 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
...
+  for (auto str : linkOptions) {
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1675]:
```suggestion
  for (const char *Str : LinkOptions) {
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 6]
```diff
@@ -1663,15 +1665,27 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
...
+    if (str != nullptr) {
+      linkOptionsStr += std::string(str) + " ";
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1677]: Can we reused applyOptionsFromImage function instead?

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+  Mock.redefine<sycl::detail::PiApiKind::piProgramCompile>(
+      redefinedProgramCompile);
+  Mock.redefine<sycl::detail::PiApiKind::piProgramLink>(redefinedProgramLink);
+
+  const sycl::device Dev = Plt.get_devices()[0];
+
+  current_link_options.clear();
+  current_compile_options.clear();
+  std::string expected_compile_options = "-cl-single-precision-constant";
+  std::string expected_link_options = "-cl-finite-math-only";
```
Comment (romanovvlad) [inline sycl/unittests/program_manager/passing_link_and_compile_options.cpp:258]: 1. Can we also test a case when there are multiple images with their own link options are merged together?
2. It would be also nice to add similar test for `sycl::build` API while we are here.

**Review by romanovvlad — CHANGES_REQUESTED** [review]

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp (hunk 7 above)
Comment (HabKaffee) [reply sycl/unittests/program_manager/passing_link_and_compile_options.cpp:258]: Yes, sure. I'll add such kind of tests.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 6 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1677]: I think we could. I'll try it.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 2 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1669]: Changed to suggested option.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 3 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1668]: Changed to suggested option.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 4 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1674]: Changed to suggested option.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 5 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1675]: Changed to suggested option.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 6 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1677]: I reused applyOptionsFromImage function and remove highlited for-cycle.

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 8]
```diff
@@ -1663,15 +1665,28 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
...
+    const std::shared_ptr<device_image_impl> &InputImpl =
+        getSyclObjImpl(DeviceImage);
+    const char *DeviceLinkOptions =
+        InputImpl->get_bin_image_ref()->getLinkOptions();
+    if (DeviceLinkOptions != nullptr) {
+      std::string TemporaryStr(DeviceLinkOptions);
+      std::optional<std::string> CompileOpts;
+      applyOptionsFromImage(*CompileOpts, TemporaryStr,
+                            *(InputImpl->get_bin_image_ref()));
+      LinkOptionsStr += TemporaryStr;
```
Comment (steffenlarsen) [inline sycl/source/detail/program_manager/program_manager.cpp:1679]: If I am reading `applyOptionsFromImage` correctly, it will append `getLinkOptions` of the supplied `RTDeviceBinaryImage` to `LinkOpts` so `TemporaryStr` would contain the linker options for each image twice. Please correct me if I am wrong.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 9]
```diff
@@ -1663,15 +1665,28 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
...
+  std::string LinkOptionsStr;
+  for (const device_image_plain &DeviceImage : DeviceImages) {
+    const std::shared_ptr<device_image_impl> &InputImpl =
+        getSyclObjImpl(DeviceImage);
+    const char *DeviceLinkOptions =
+        InputImpl->get_bin_image_ref()->getLinkOptions();
+    if (DeviceLinkOptions != nullptr) {
+      std::string TemporaryStr(DeviceLinkOptions);
+      std::optional<std::string> CompileOpts;
+      applyOptionsFromImage(*CompileOpts, TemporaryStr,
```
Comment (steffenlarsen) [inline sycl/source/detail/program_manager/program_manager.cpp:1677]: It is [undefined behavior](https://en.cppreference.com/w/cpp/utility/optional/operator*) to use the `*` operator on an `std::optional` that has no value. Since `CompileOpts` is [created without a value](https://en.cppreference.com/w/cpp/utility/optional/optional), `*CompileOpts` here is undefined behavior.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 8 above)
Comment (alexbatashev) [reply sycl/source/detail/program_manager/program_manager.cpp:1679]: You’re right, but it should not be a problem. Compilers can generally handle that. On the contrary, parsing options on rt side is hard: `-cc1 -fsycl-device-only`

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 9 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1677]: Could I set it to an empty string to define its behaviour? I mean change it to `std::optional<std::string> CompileOpts("");`

Comment (steffenlarsen) [reply sycl/source/detail/program_manager/program_manager.cpp]: You could, but I don't see a benefit to that over just having `std::string CompileOpts`.

Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp]: Yes, you are right.

Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp]: Changed to `std::string`

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 10]
```diff
@@ -1698,15 +1700,22 @@ ProgramManager::link(const std::vector<device_image_plain> &DeviceImages,
   for (const device &Dev : Devs)
     PIDevices.push_back(getSyclObjImpl(Dev)->getHandleRef());
 
+  std::string LinkOptionsStr;
+  for (const device_image_plain &DeviceImage : DeviceImages) {
+    const std::shared_ptr<device_image_impl> &InputImpl =
+        getSyclObjImpl(DeviceImage);
+    std::string CompileOpts;
+    applyOptionsFromImage(CompileOpts, LinkOptionsStr,
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1708]: Minor.
Can we customize `applyOptionsFromImage` to avoid filling `CompileOpts` when not requested? I think we could, split the function to `applyCompileOptionsFromImage` and `applyLinkOptionsFromImage` which would handle compile and link options separately.

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+
+#include <helpers/CommonRedefinitions.hpp>
+#include <helpers/PiImage.hpp>
+#include <helpers/PiMock.hpp>
+
+#include <gtest/gtest.h>
+
+std::string current_link_options, current_compile_options;
+
+class EAMTestKernel1;
```
Comment (romanovvlad) [inline sycl/unittests/program_manager/passing_link_and_compile_options.cpp:19]: Just curious, what does `EAM` stand for ?

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+  setupDefaultMockAPIs(Mock);
+  Mock.redefine<sycl::detail::PiApiKind::piProgramCompile>(
+      redefinedProgramCompile);
+  Mock.redefine<sycl::detail::PiApiKind::piProgramLink>(redefinedProgramLink);
+
+  const sycl::device Dev = Plt.get_devices()[0];
+  // Check devices' compile & link options separately
+  current_link_options.clear();
+  current_compile_options.clear();
+  std::string expected_compile_options1 = "", expected_compile_options2 = "";
```
Comment (romanovvlad) [inline sycl/unittests/program_manager/passing_link_and_compile_options.cpp:168]: Do we really need several strings here? It seems the test is about empty options, so one "expected" string would be enough?

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp [hunk 13]
```diff
@@ -0,0 +1,529 @@
...
+  current_compile_options.clear();
+  static sycl::unittest::PiImageArray<2> DevImageArray[] = {&DevImage1,
+                                                            &DevImage2};
+  sycl::kernel_bundle KernelBundle =
+      sycl::get_kernel_bundle<sycl::bundle_state::input>(
+          Ctx, {Dev}, {KernelID_1, KernelID_2});
+  auto BundleObj = sycl::compile(KernelBundle);
+  sycl::link(BundleObj);
+  EXPECT_EQ("", current_link_options);
+  EXPECT_EQ("", current_compile_options);
```
Comment (romanovvlad) [inline sycl/unittests/program_manager/passing_link_and_compile_options.cpp:209]: Can we check against "expected" string variable for consistency?

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp [hunk 14]
```diff
@@ -0,0 +1,529 @@
...
+  // empty str + not empty str
+  DevImage1 = generateEAMTestKernel1Image<EAMTestKernel1>(
+      expected_compile_options1_1, expected_link_options);
+  current_link_options.clear();
+  current_compile_options.clear();
+  static sycl::unittest::PiImageArray<2> DevImageArray1[] = {&DevImage1,
+                                                             &DevImage2};
+
+  BundleObj = sycl::compile(KernelBundle);
+  sycl::link(BundleObj);
```
Comment (romanovvlad) [inline sycl/unittests/program_manager/passing_link_and_compile_options.cpp:294]: Could you please tell why linking is done here? It seems the test doesn't check for link options?

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp [hunk 15]
```diff
@@ -0,0 +1,529 @@
...
+  static sycl::unittest::PiImageArray<2> DevImageArray2[] = {&DevImage1,
+                                                             &DevImage2};
+
+  BundleObj = sycl::compile(KernelBundle);
+  sycl::link(BundleObj);
+  EXPECT_EQ("", current_link_options);
+  EXPECT_EQ(expected_compile_options1, current_compile_options);
+}
+
+TEST(Link_Compile_Options, compile_link_Options_Test_only_link_options) {
```
Comment (romanovvlad) [inline sycl/unittests/program_manager/passing_link_and_compile_options.cpp:313]: Minor.
I'm not sure it makes a lot of sense to test compile and link options individually. I believe `link_compile_options_all_options` test covers both good enough.

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp (hunk 7 above)
Comment (romanovvlad) [reply sycl/unittests/program_manager/passing_link_and_compile_options.cpp:515]: Thanks

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp (hunk 11 above)
Comment (HabKaffee) [reply sycl/unittests/program_manager/passing_link_and_compile_options.cpp:19]: That kernel was used from another test.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 10 above)
Comment (HabKaffee) [reply sycl/source/detail/program_manager/program_manager.cpp:1708]: I think we could, because I also find it more comfortable. I'll implement it.

**Comment by HabKaffee** [comment]
@romanovvlad Should I convert this PR to draft due to last requests?

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp (hunk 12 above)
Comment (HabKaffee) [reply sycl/unittests/program_manager/passing_link_and_compile_options.cpp:168]: You are right. Using two empty string is unnecessary here.

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp (hunk 15 above)
Comment (HabKaffee) [reply sycl/unittests/program_manager/passing_link_and_compile_options.cpp:313]: I got your point. Now here would be 2 tests: with both empty options and with both options filled.

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp (hunk 13 above)
Comment (HabKaffee) [reply sycl/unittests/program_manager/passing_link_and_compile_options.cpp:209]: Yes, of course. I'll change it

File: sycl/unittests/program_manager/passing_link_and_compile_options.cpp (hunk 14 above)
Comment (HabKaffee) [reply sycl/unittests/program_manager/passing_link_and_compile_options.cpp:294]: Yes, this test is checking only compile options. I'm rewriting tests now and I'm going to remove all unnecessary actions.

**Comment by romanovvlad** [comment]
> @romanovvlad Should I convert this PR to draft due to last requests?

It's up to you. Do not think this is necessary.

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
I have a question. What should I do with program_manager/program_manager.cpp:987 which is ` LinkDeviceLibs = false;`? If I remove that line tests in intel/llvm-test-suite are failing. (approximately 17 tests)

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 16]
```diff
@@ -981,9 +995,12 @@ ProgramManager::ProgramPtr ProgramManager::build(
 
   const detail::plugin &Plugin = Context->getPlugin();
   if (LinkPrograms.empty() && !ForceLink) {
+    std::string Options = LinkOptions.empty()
+                              ? CompileOptions
+                              : (CompileOptions + " " + LinkOptions);
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1000]:
```suggestion
    const std::string &Options = LinkOptions.empty()
                              ? CompileOptions
                              : (CompileOptions + " " + LinkOptions);
```

**Review by romanovvlad — DISMISSED** [review]

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Review by romanovvlad — DISMISSED** [review]

**Review by romanovvlad — DISMISSED** [review]

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Review by steffenlarsen — DISMISSED** [review]
Only a minor nit, otherwise LGTM!

**Comment by HabKaffee** [comment]
@steffenlarsen @romanovvlad Could you, please, reapprove this PR?

**Review by steffenlarsen — APPROVED** [review]

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
I have resolved merge conflict between my branch and intel/sycl. @steffenlarsen @romanovvlad could you, please, reapprove this PR?

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Review by steffenlarsen — APPROVED** [review]

**Comment by HabKaffee** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/806

**Comment by HabKaffee** [comment]
All check, which are failing, were fixed by https://github.com/intel/llvm-test-suite/pull/806 .

**Comment by bader** [comment]
@vladimirlaz, could you merge this PR, please?

**Comment by vladimirlaz** [comment]
> @vladimirlaz, could you merge this PR, please?

will do after 6PM MSK

**Review by vladimirlaz — APPROVED** [review]

**Comment by vladimirlaz** [comment]
@HabKaffee it looks like we have post-commit build failure with the patch

**Comment by bader** [comment]
Is it caused by this patch or https://github.com/intel/llvm/commit/87f60f6e7c3f07fc816fd1e0f667c98e881f4933?

**Comment by HabKaffee** [comment]
@vladimirlaz I see, that https://github.com/intel/llvm/commit/87f60f6e7c3f07fc816fd1e0f667c98e881f4933 which was before my commit failed

**Comment by vladimirlaz** [comment]
Sorry, you are right...
